File: python/kocab/line.py
from datetime import datetime

from . import status


class Line:
    def __init__(self, time: datetime, status: status.Status):
        self.time = time
        self.status = status

    def __str__(self):
        return str(self.time) + " " + str(self.status)

    def set_default(self):
        self.time = datetime(1, 1, 1, 0, 0, 0, 0)
        self.status = status.get_status_by_id(status.DEFAULT_ID)

    def is_zero(self) -> bool:
        return self.time.hour == 0 and self.time.minute == 0 and self.time.second == 0 and self.time.microsecond == 0


def new() -> Line:
    return Line(datetime(1, 1, 1, 0, 0, 0, 0), status.get_status_by_id(status.DEFAULT_ID))


def parse(raw_time: str, raw_id: str) -> Line:
    milisecons = (int(raw_time) % 1000) * 1000
    seconds = (int(raw_time) // 1000) % 60
    minutes = (int(raw_time) // 1000 // 60) % 60
    hours = (int(raw_time) // 1000 // 60 // 60) % 24

    line_time = datetime(1, 1, 1, hours, minutes, seconds, milisecons)
    line_status = status.parse_raw_status(raw_id)

    return Line(line_time, line_status)


def parse_countdown(raw_time: str) -> Line:
    milisecons = (int(raw_time) % 1000) * 1000
    seconds = (int(raw_time) // 1000) % 60
    minutes = (int(raw_time) // 1000 // 60) % 60
    hours = (int(raw_time) // 1000 // 60 // 60) % 24

    line_time = datetime(1, 1, 1, hours, minutes, seconds, milisecons)
    result = Line(line_time, status.get_status_by_id(status.UNDEFINED_ID))

    if result.is_zero():
        result.status = status.get_status_by_id(status.STOP_ID)
    else:
        result.status = status.get_status_by_id(status.RUN_ID)

    return result

File: app/timers/line/line.go
package line

import (
	"fmt"
	"strconv"
	"time"

	"thomasparsley.cz/firesport-timer/timers/status"
)

type Line struct {
	Time   time.Time     `json:"time"`
	Status status.Status `json:"status"`
}

func (l Line) IsZero() bool {
	return l.Time.Minute() == 0 &&
		l.Time.Second() == 0 &&
		l.Time.Nanosecond() == 0
}
func (l *Line) SetDefault() {
	newLine := New()
	l = &newLine
}

func New() Line {
	return Line{
		Time:   time.Date(1, 1, 1, 0, 0, 0, 0, time.Local),
		Status: status.New(),
	}
}

func Parse(rawTime string, rawID string) (Line, error) {
	timeInt, err := strconv.Atoi(rawTime)
	if err != nil {
		return Line{}, err
	}

	t := time.Date(1, 1, 1, 0, 0, 0, timeInt*1000000, time.Local)

	s, err := status.ParseRaw(rawID)
	if err != nil {
		return Line{}, err
	}

	return Line{
		Time:   t,
		Status: s,
	}, nil
}

func ParseCountdown(rawTime string) (Line, error) {
	result, err := Parse(rawTime, fmt.Sprintf("%d", status.UndefinedID))
	if err != nil {
		return Line{}, err
	}

	if result.IsZero() {
		result.Status = status.GetByID(status.StopID)
	} else {
		result.Status = status.GetByID(status.RunID)
	}

	return result, nil
}
